library/alloc/src/raw_vec.rs

```diff
@@ -394,6 +394,24 @@ impl<T, A: Allocator> RawVec<T, A> {
         let cap = cmp::max(self.cap * 2, required_cap);
         let cap = cmp::max(Self::MIN_NON_ZERO_CAP, cap);
 
+        let alloc_size = cap.checked_mul(mem::align_of::<T>()).ok_or(CapacityOverflow)?;
+
+        // Since memory allocators tend to use power of two sized bins, find the
+        // bin size we will fall into.
+        let bin_size = usize::MAX >> (alloc_size.leading_zeros() - 1);
+
+        // Leave some room for allocators that add fixed overhead (usually one pointer-size)
+        // FIXME: maybe two pointers to be on the safe side? It could potentially be platform-dependent.
```

Comment:
Hmm, this is a good point.  This change would massively backfire if it underestimates the overhead, as the underlying allocator would then need to massively overallocate basically every time, instead of the sometimes you'd get today depending on the original capacity.

So maybe this should be very conservative in the space it leaves for the allocator overhead?

Reply:
Thinking about it again, though, is leaving any space at all pessimizing good allocators in favour of mediocre ones?

One big advantage of rust's allocator APIs is that deallocation passes the size.  So, for example, if an allocator allocates dedicated (maybe-large) pages for 32-byte allocations, they don't necessarily need any per-allocation overhead to do that, as they can tell by the deallocation size that it's a page of 32-byte allocations, and could have per-page overhead instead.

Reply:
Generally, I think that any allocator that uses most than just a single pointer is meh, and is not worth pessimizing. 

But yeah... unfortunately this optimization is assuming implementation details that are invisible. Hopefully benchmarks will tell if any platform degraded, and we could always use conditional compilation or something to tweak it.

Reply:
There's another interesting case to consider: the windows `System` allocator has to implement alignment on top of allocation which does not support custom alignment. This means it allocates `align` extra bytes.

Yes, Rust's allocation API allows the use of allocators which don't have to store metadata in allocations, but in practice, most allocators will want to support the C allocation API.

Reply:
> alignment on top of allocation which does not support custom alignment. This means it allocates align extra bytes.

@CAD97 I don't understand.

Reply:
https://github.com/rust-lang/rust/blob/dec689432fac6720b2f18101ac28a21add98b1b8/library/std/src/sys/windows/alloc.rs#L129-L178

Roughly:

```rust
fn win32::HeapAlloc(hHeap: win32::HANDLE, dwFlags: win32::DWORD, dwBytes: win32::SIZE_T) -> win32::LPVOID;
fn win32::HeapFree (hHeam: win32::HANDLE, dwFlags: win32::DWORD, lpMem: win32::LPVOID);

fn HeapAlignedAlloc(heap: win32::HANDLE, flags: win32::DWORD, size: usize, align: usize) -> *mut u8 {
    if align <= win32::MEMORY_ALLOCATION_ALIGNMENT {
        return win32::HeapAlloc(heap, flags, size) as *mut u8;
    }
    let ptr = win32::HeapAlloc(heap, flags, size + align);
    let aligned = (ptr as *mut u8).add(1).align_to(align);
    *(aligned as *mut win32::LPVOID).sub(1) = ptr;
    aligned
}

fn HeapAlignedFree(heap: win32::HANDLE, flags: win32::DWORD, ptr: *mut u8, size: usize, align: usize) {
    if align <= win32::MEMORY_ALLOCATION_ALIGNMENT {
        win32::HeapFree(heap, flags, ptr as _)
    } else {
        win32::HeapFree(heap, flags, *(ptr as *mut win32::LPVOID).sub(1))
    }
}
```

Windows doesn't provide a way to `HeapAlloc` at anything other than `MEMORY_ALLOCATION_ALIGNMENT`. (Windows does offer `_alligned_malloc`, `_aligned_realloc`, `_aligned_recalloc`, and `_aligned_free`, but strangely not `_aligned_calloc`.) `std::alloc::System` is documented to use `HeapAlloc` on Windows.

Reply:
> There's another interesting case to consider: the windows `System` allocator has to implement alignment on top of allocation which does not support custom alignment. This means it allocates `align` extra bytes.

Wouldn't memory allocator request much, much larger chunks of memory from the system at once, and then use them for fine-grained user-allocations individually? So any system level alignment issues would generally have much smaller overhead?

Reply:
An actual new allocator implementation, yes, typically. I'm specifically referring to `System` or other Rust wrappers of allocators which don't provide an aligned allocation interface, which need to do this technique on a per-allocation basis.

While this may be only the Windows system allocator in practice, this is still the default allocator that most Rust programs compiled for Windows will be using, so it's an important case to consider.

Reply:
I kind of understand, but not enough to come up with a solution. Worst case we could just have a different implementation for Windows maybe?